### PR TITLE
Add signal

### PR DESCRIPTION
**Pull request by ian-r-rose**
Okay, hopefully I have this right.  I am trying to rewrite some things to take advantage of the new signals/slots mechanism.  As I understand from the section Wolfgang wrote in the manual, we are to add new listener signals in the core of Aspect as needed?

This particular signal is for hacks where you might want to edit certain of the parameters on the fly.  The test I have added changes one of the temperature boundary conditions from fixed-temperature to no-flux after a few timesteps.

### CODE REVIEW

File: include/aspect/simulator_signals.h [hunk 1]
```diff
@@ -70,6 +70,20 @@ namespace aspect
                                   ConstraintMatrix &)>  post_constraints_creation;
 
     /**
+    * A signal that is called before setting up the degrees of
+    * freedom for the simulator. This allows for editing of the parameters
+    * struct  on the fly (such as changing boundary conditions) to give
+    * Aspect different behaviour in mid-run than it otherwise would have.
```
Comment (bangerth) [inline include/aspect/simulator_signals.h]: You're American -- I'm surprised you spell behavior as you do :-)

There's also a double space on the line before.

File: source/simulator/core.cc [hunk 2]
```diff
@@ -989,6 +989,8 @@ namespace aspect
   template <int dim>
   void Simulator<dim>::setup_dofs ()
   {
+    signals.edit_parameters(*this, parameters);
+
```
Comment (bangerth) [inline source/simulator/core.cc]: I wonder if this is the right place to call the signal. How about at the end of `parse_parameters()`? Or what was your reasoning to call it here?

File: include/aspect/simulator_signals.h (hunk 1 above)
Comment (ian-r-rose) [inline include/aspect/simulator_signals.h]: Hmm, too much BBC :-)

File: tests/edit_parameters.cc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  template <int dim>
+  void change_boundary_condition (const SimulatorAccess<dim> &simulator_access,
+                                  Parameters<dim> &parameters)
+  {
+    types::boundary_id bottom = 2; //bottom boundary indicator for 2D box
+
+    if ( simulator_access.get_timestep_number() >= switch_step && !switched )
+      {
```
Comment (bangerth) [inline tests/edit_parameters.cc]: ah, I see -- you want to have the signal called at the beginning of each timestep. Isn't there a `start_timestep()` or `begin_timestep()` function or similar from which this signal could be called?

I think it would be worthwhile documenting in the description of the signal where and when the signal is triggered.

**Comment by bangerth** [comment]
Looks generally ok.

And yes, the idea is to liberally invent signal wherever you find a need for them.

File: source/simulator/core.cc (hunk 2 above)
Comment (ian-r-rose) [inline source/simulator/core.cc]: There is a reasoning: doing it at the end of ```parse_parameters()``` means it is only called once. If I want to do some editing at other times, it needs to be in the main loop. 

I put it in  ```setup_dofs()``` because in many cases edits to the parameters only take effect when this function is called.  For instance, setting up the boundary conditions happens here, so in the use case that I've tested, this is the most appropriate place for it.

This is certainly a hack, but my understanding that this is what the signals/slots are for?

File: tests/edit_parameters.cc (hunk 3 above)
Comment (ian-r-rose) [inline tests/edit_parameters.cc]: Yes, in fact, in my initial experiments I had this in ```start_timestep()```.  I moved it to ```setup_dofs()``` for the reason outlined above, but I am not particularly wedded to that, and would be happy to put it back in ```start_timestep()```

File: source/simulator/core.cc (hunk 2 above)
Comment (bangerth) [inline source/simulator/core.cc]: But isn't setup_dofs only called when the mesh has been changed?

Comment (ian-r-rose) [inline source/simulator/core.cc]: Yeah... for now I am putting it back in ```start_timestep()```.  It doesn't matter much from my perspective, and that will cover more potential use cases.

Comment (bangerth) [inline source/simulator/core.cc]: I don't care one way or the other, but I think if you call it `edit_parameters` then either once or each time step seems reasonable. If you have good reasons to do it at the start of `setup_dofs()`, then call it `pre_setup_dofs()` :-)

File: include/aspect/simulator_signals.h [hunk 4]
```diff
@@ -70,6 +70,24 @@ namespace aspect
                                   ConstraintMatrix &)>  post_constraints_creation;
 
     /**
+    * A signal that is called at the start of every timestep.
```
Comment (tjhei) [inline include/aspect/simulator_signals.h]: At every timestep or only when setup_dofs is called?

Comment (ian-r-rose) [inline include/aspect/simulator_signals.h]: Right you are -- I moved the signal, and the docs became out of date...

**Comment by bangerth** [comment]
The name is better now, thanks!

This is good to merge. The next person to come by this PR after the tests have successfully run, please merge!

/run-tests
